rfcs/text/0017_url_service.md

```diff
@@ -0,0 +1,585 @@
+
+We have identified the following limitations in the current implementation of
+the URL Generator Service:
+
+1. URL generators are available only on the client-side. There is no way to use
+   them together with short URLs.
+   1. __Will do:__ We will implement locators also on the server-side
+      (they will be available in both environments) and we will combine them
+      with the Short URL Service.
+1. URL generators are not exposed externally, thus Cloud and Support cannot use
+   them to generate deep links into Kibana.
+   1. __Will do:__ We will expose HTTP endpoints on the server-side and the
+      "redirect" app on the client-side which external users will be able to use
+      to deep link into Kibana using locators.
+
+
+## Limitations of the architecture
+
+One major reason we want to "refresh" the Short URL Service and the URL
+Generator Service is their architecture.
+
+Currently, the Short URL Service is implemented on top of the `url` type saved
+object on the server-side. However, it only exposes the
+HTTP endpoints, it does not expose any API on the server for the server-side
+plugins to consume; on the client-side there is no plugin API either, developers
+need to manually execute HTTP requests. 
+
+The URL Generator Service is only available on the client-side, there is no way
+to use it on the server-side, yet we already have use cases (for example ML
+team) where a server-side plugin wants to use a URL generator.
+
+![Current Short URL Service and URL Generator Service architecture](../images/url_service/old_architecture.png)
+
+The current architecture does not allow both services to be conveniently used,
+also as they are implemented in different locations, they are disjointed&mdash;
+we cannot create a short URL using an URL generator.
+
+
+# Detailed design
+
+In general we will try to provide as much as possible the same API on the
+server-side and the client-side.
+
+
+## High level architecture
+
+Below diagram shows the proposed architecture of the URL Service.
+
+![URL Service architecture](../images/url_service/new_architecture.png)
+
+
+## Plugin contracts
+
+The aim is to provide developers the same experience on the server and browser.
+
+Below are preliminary interfaces of the new URL Service. `IUrlService` will be
+a shared interface defined in `/common` folder shared across server and browser.
+This will allow us to provide users a common API interface on the server and
+browser, wherever they choose to use the URL Service:
+
+```ts
+/**
+ * Common URL Service client interface for the server-side and the client-side.
+ */
+interface IUrlService {
+  locators: ILocatorClient;
+  shortUrls: IShortUrlClient;
+}
+```
+
+
+### Locators
+
+The locator business logic will be contained in `ILocatorClient` client and will
+provide two main functionalities:
+
+1. It will provide a facility to create locators.
+1. It will also be a registry of locators, every newly created locator is
```

Comment:
lets mention when this registry should/shouldnt be used:

`the registry should never be used when locator ID is known at the compile time, but is resolved only for usecases where we only know ID of locator at the runtime.`

Reply:
Added you comment.